### PR TITLE
DEPR: rename to _consolidate and create deprecation warning

### LINKED ISSUE #15483: DEPR: .consolidate() should be non-public
no user reason to use this, and will eventually disappear in pandas 2.0 anyhow.

I would rename to ``_consolidate()`` and then provide a deprecation warning for ``.consolidate()``. Would need to change some internal code to use the renamed version.

xref https://github.com/pandas-dev/pandas/issues/15482#issuecomment-281986233

### PR DESCRIPTION
**Pull request by GuessWhoSamFoo**
- [x] closes #15483 
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Test output:
```
running: pytest --skip-slow --skip-network pandas
============================= test session starts ==============================
platform linux2 -- Python 2.7.13, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /home/sfoo/pandas, inifile: setup.cfg
collected 11882 items / 4 skipped

 10147 passed, 1662 skipped, 77 xfailed, 2084 pytest-warnings in 407.18 seconds  
```

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@GuessWhoSamFoo We would like to have it the other way around. Have the deprecation warning in `consolidate` and not in `_consolidate` (so this last one can be used internally without warning). So you can leave the existing `consolidate` intact like it is now, apart from renaming it to `_consolidate`, and then add a new `consolidate` that has the deprecation warning and just calls `_consolidate` (not adding a new `_consolidate` like you did now). 
Further, you will also have to replace internal usage in the pandas codebase of `consolidate` to `_consolidate` to avoid raising the deprecation warning (that is the reason for the test failures).

**Comment by GuessWhoSamFoo** [comment]
```running: pytest --skip-slow --skip-network pandas
==================================== test session starts =====================================
platform linux2 -- Python 2.7.13, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /home/sfoo/pandas, inifile: setup.cfg
collected 11882 items / 4 skipped 

======= 10147 passed, 1662 skipped, 77 xfailed, 2084 pytest-warnings in 419.22 seconds =======
```
The original consolidate has been left intact except for a line taken out of the docstring. Then I've changed instances of `consolidate` to `_consolidate` (even in tests) but still have the same number of failed tests/warnings. Am I missing something here?

File: pandas/io/pytables.py [hunk 1]
```diff
@@ -2861,7 +2861,7 @@ def write(self, obj, **kwargs):
         super(BlockManagerFixed, self).write(obj, **kwargs)
         data = obj._data
         if not data.is_consolidated():
-            data = data.consolidate()
+            data = data._consolidate()
```
Comment (jorisvandenbossche) [inline pandas/io/pytables.py]: Here, `data` is not a DataFrame, but a BlockManager, so this should be left as `consolidate`

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -484,6 +484,7 @@ Deprecations
 - ``DataFrame.astype()`` has deprecated the ``raise_on_error`` parameter in favor of ``errors`` (:issue:`14878`)
 - ``Series.sortlevel`` and ``DataFrame.sortlevel`` have been deprecated in favor of ``Series.sort_index`` and ``DataFrame.sort_index`` (:issue:`15099`)
 - importing ``concat`` from ``pandas.tools.merge`` has been deprecated in favor of imports from the ``pandas`` namespace. This should only affect explict imports (:issue:`15358`)
+- ``DataFrame.consolidate()`` has been deprecated. (:issue:`15483`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: say ``Series/DataFrame/Panel.consolidate()`` it affects everything

say has been deprecated as a public method as

File: pandas/core/generic.py [hunk 3]
```diff
@@ -2874,11 +2874,12 @@ def f():
 
         self._protect_consolidate(f)
 
-    def consolidate(self, inplace=False):
+    def _consolidate(self, inplace=False):
         """
+        DEPRECATED:
```
Comment (jreback) [inline pandas/core/generic.py]: this is not deprecated though.

can you add a doc-string

File: pandas/core/generic.py [hunk 4]
```diff
@@ -2897,6 +2898,12 @@ def consolidate(self, inplace=False):
             cons_data = self._protect_consolidate(f)
             return self._constructor(cons_data).__finalize__(self)
 
+    def consolidate(self, inplace=False):
+        # 15483
```
Comment (jreback) [inline pandas/core/generic.py]: here is where the DEPRECATED goes (in the doc-string)

File: pandas/tests/frame/test_block_internals.py [hunk 5]
```diff
@@ -40,17 +40,17 @@ def test_cast_internals(self):
 
     def test_consolidate(self):
```
Comment (jreback) [inline pandas/tests/frame/test_block_internals.py]: can you add a single test (test_consolidate_deprecation), which uses ``assert_produces_warning(FutureWarning)``

File: pandas/core/generic.py [hunk 6]
```diff
@@ -2897,6 +2898,12 @@ def consolidate(self, inplace=False):
...
+        warnings.warn("consolidate is deprecated and will be removed in a "
+                      "future release.", DeprecationWarning)
```
Comment (jreback) [inline pandas/core/generic.py]: use FutureWarning

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by GuessWhoSamFoo** [comment]
Squashed the last two commits with the changes from the feedback. Thank you for the guidance!

**Comment by jorisvandenbossche** [comment]
@GuessWhoSamFoo Thank you!

**Pull request merged**